lib/gc/Transforms/Microkernel/MicrokernelInvariantCodeMotion.cpp

```diff
@@ -0,0 +1,434 @@
+      return rewriter.notifyMatchFailure(op, "Not call to BRGEMM tilecfg");
+    auto opInfoIter = structInfo.find(op);
+    if (opInfoIter == structInfo.end()) {
+      return rewriter.notifyMatchFailure(op, "Cannot find structInfo for Op");
+    }
+    auto &opStructInfo = opInfoIter->second;
+
+    // Don't hoist if max invariant scope is itself to reduce
+    // unnecessary movement
+    if (opStructInfo.maxInvariantScope == op) {
+      return rewriter.notifyMatchFailure(op, "No need to hoist");
+    }
+    rewriter.moveOpBefore(op, opStructInfo.maxInvariantScope);
+    // Avoid being hoisted again
+    opStructInfo.maxInvariantScope = op;
+    return success();
+  }
+};
+
+static void markScopeAsReleased(OpStructInfoMap &structInfo, Operation *op) {
+  auto iter = structInfo.find(op);
+  assert(iter != structInfo.end());
+  // Don't mark BrgemmCallOps
+  if (getBrgemmCallType(op) != BrgemmCallType::INAPPLICABLE)
+    return;
+  iter->second.hasTilereleased = true;
+
+  for (auto ch : iter->second.child) {
+    markScopeAsReleased(structInfo, ch);
+  }
+}
+
+class BrgemmTilereleaseRewriter : public OpRewritePattern<func::CallOp> {
+private:
+  OpStructInfoMap &structInfo;
+
+public:
+  using OpRewritePattern<func::CallOp>::OpRewritePattern;
+
+  BrgemmTilereleaseRewriter(MLIRContext *context, OpStructInfoMap &si)
+      : OpRewritePattern(context), structInfo{si} {}
+  LogicalResult matchAndRewrite(func::CallOp op,
+                                PatternRewriter &rewriter) const final {
+    ModuleOp module = op->template getParentOfType<ModuleOp>();
+
+    StringAttr callee = op.getCalleeAttr().getAttr();
+    if (!module.lookupSymbol(callee))
+      return rewriter.notifyMatchFailure(op,
+                                         "Invalid CallOp to unknown callee");
+    if (callee !=
+        StringAttr::get(rewriter.getContext(), DNNL_BRGEMM_TILERELEASE_NAME))
+      return rewriter.notifyMatchFailure(op, "Not call to BRGEMM tilerelease");
+
+    auto opInfoIter = structInfo.find(op);
+    if (opInfoIter == structInfo.end()) {
+      return rewriter.notifyMatchFailure(op, "Cannot find structInfo for Op");
+    }
+    auto &opStructInfo = opInfoIter->second;
+    auto targetInfoIter = structInfo.find(opStructInfo.maxInvariantScope);
+    assert(opStructInfo.maxInvariantScope);
+    // Don't hoist if max invariant scope is itself to reduce
+    // unnecessary movement
+    if (opStructInfo.maxInvariantScope == op) {
+      return rewriter.notifyMatchFailure(op, "No need to hoist");
+    }
+    assert(targetInfoIter != structInfo.end());
+    // move last tilerelease to end of contextRoot, and remove all
+    // others
+    if (targetInfoIter->second.hasTilereleased) {
+      rewriter.eraseOp(op);
+    } else {
+      // auto region = opStructInfo.maxInvariantScope->getRegion(0);
+      // auto block = &region.getBlocks().front();
+      // auto enditer = block->end();
+      // rewriter.moveOpBefore(op, block, enditer);
```

Comment:
Are these useless? Or do you have any purpose to keep them?

Reply:
Useless comments removed.